### PR TITLE
feat(driver): Checkpoint Sync

### LINKED ISSUE #41: feat(driver): fast sync
Follow chain head using derived data, but perform the initial sync by reading the L2 output oracle on L1 to find the current head hash and use the engine's built in sync mechanism to sync to that head.

### PR DESCRIPTION
**Pull request by merklefruit**
Closes #41

This PR implements syncing from a checkpoint L2 block hash.

## Syncing strategy:

1. By default, find the L2 block corresponding to the epoch boundary (with `seq_index = 0`) and use its hash as checkpoint. Otherwise, give the user the option to provide a checkpoint manually via `--checkpoint-hash` cli flag
2. Build the entire payload of that block using a trusted L2 RPC
3. Send a `newPayloadV1` request to the execution layer for that block
4. Send a `forkchoice_updated` request with the checkpoint block hash
5. Syncing should start at this point. Monitor its progress via `eth_syncing` or `eth_blockNumber`
6. Once done syncing the execution layer, fetch the new head info and start the driver "normally"

## TODO:

- [x] add `checkpoint_hash` cli arg: the block hash to start sync from. (will be made optional for fast sync)
- [x] add `l2_trusted_rpc_url` cli arg: the rpc to pull the initial payload from, in order to start syncing
- [x] parse the checkpoint block hash into an execution payload
- [x] send that payload to the execution layer via `new_payload`
- [x] call `forkchoice_updated` with the checkpoint hash
- [x] wait for the execution layer to sync via `eth_syncing` or `eth_blockNumber`
- [x] add scripting to perform checkpoint sync in `start-op-geth.sh`
- [x] start the normal sync process by restarting sync from new head
- [x] add ability to fetch the checkpoint automatically from a trusted rpc as the latest epoch boundary

## CHANGELOG

- Added a "runner" module, which follows the builder pattern and implements all the different syncing functions
- Magi binary: added `Checkpoint` SyncMode, added `checkpoint_hash` and `l2_trusted_rpc_url` cli args.
- chain config: added `l2_output_oracle` (not used yet, needs consensus layer p2p), `l1_block` and `sequencer_fee_vault` contract addresses for both opt and base goerli
- driver types: added a method to turn an L2 block into HeadInfo struct
- engine: added a method to create a forkchoice from a single hash for convenience
- engine payload: added a method to create a new ExecutionPayload from a block hash. This process required me to patch `ethers-rs` in order to add support for the Optimism deposited transaction type, along with its RLP encoding.
- Some minor changes to tests and docs.
- slightly refactored the .env.default file to hopefully be more clear for new users

### CODE REVIEW

**Comment by ncitron** [comment]
This seems to be going in the right direction. I'm curious why you chose to fetch the output oracle address via the portal instead of fetching it via the config?

**Comment by merklefruit** [comment]
> fetch the output oracle address via the portal instead of fetching it via the config

What other config contract holds the oracle address? Or do you mean saving the oracle address in the `ChainConfig` struct?

I just took inspiration from what op-node does [here](https://github.com/ethereum-optimism/optimism/blob/develop/op-node/withdrawals/utils.go#L37) for fetching from the portal

**Comment by merklefruit** [comment]
@ncitron I have a few more questions to wrap my head around the implementation. Please bear with me 😅

1. now that the full sync starts from the last finalized block by default (instead of genesis) should we add the loop to check `eth_syncing` status to the full-sync process as well as the fast-sync?

2. when initializing the driver from a checkpoint hash, I try to fetch the head info from the L2 rpc, but since op-geth is not synced to that block yet, it will return empty blocks or errors. How can I fetch the head info from an L2 block that I don't have yet?

3. I noticed that we do the following loop in the `Driver::start()` method:
    ```rust
    async fn await_engine_ready(&self) {
            while !self.engine_driver.engine_ready().await {
                self.check_shutdown().await;
                sleep(Duration::from_secs(1)).await;
            }
        }
    ```
    what is the reason for the loop? Shouldn't this request be sent only once normally?

**Comment by ncitron** [comment]
1. Nope, during full sync, `eth_syncing` should always be false since the head is only advanced one block at a time.
2. We shouldn't need the `HeadInfo` to start it off. If we call `engine.forkchoice_updated` and pass it the checkpoint hash for the finalized/safe/unsafe blocks, the engine should start syncing on its own to that block using its p2p network. Once this sync is done, we now should have the checkpoint block in our engine, so we can then derive the `HeadInfo` from that.
3. This just has the effect of preventing Magi from beginning the sync process until it successfully connects the execution engine. If the engine hasn't started up yet, this prevents Magi from crashing when it starts to sync and errors on every engine api request.

**Comment by merklefruit** [comment]
@ncitron Thanks, I now undertand the flow of calling the execution client to sync to an user-specified checkpoint block, however I am encountering an issue with the `forkchoice_updated` call. Whenever I send a request with empty attributes and a forkchoiceState object created like so:

```rust
let checkpoint_hash = "0xc2794a16acacd9f7670379ffd12b6968ff98e2a602f57d7d1f880220aa5a4973".parse()?;
let forkchoice = ForkChoiceState {
  head_block_hash: checkpoint_hash,
  safe_block_hash: checkpoint_hash,
  finalized_block_hash: checkpoint hash
}

engine_api.forkchoice_updated(forkchoice, None).await?;
```
(where checkpoint_hash is provided via the CLI with a value of a op-goerli block hash such as [this one](https://goerli-optimism.etherscan.io/block/8453214))

op-geth logs the warning: "Forkchoice requested unknown head" and the syncing status appears to be false.

By [looking at op-geth](https://github.com/ethereum-optimism/op-geth/blob/optimism/eth/catalyst/api.go#L240), this error only occurs when our client cannot resolve the header from its cache, which makes sense since we are trying to sync it up to a block which it doesn't have yet.

Do you have any suggestions on how to tackle this issue?

**Comment by ncitron** [comment]
Ok so I've looked a bit into the implementation of op-geth, and it seems that they throw a bit of a wrench into our plans.

For some reason, it seems op-geth will only let us begin the sync process if it has seen the payload from the `newPayloadV1` api call. This means for our checkpoint block we will need to reconstruct and entire payload for it. To do this, we probably want to use a trusted L2 rpc for our sync (btw I think we should rename this checkpoint sync instead of fast sync, because ethereum consensus layers do something very similar and call it this).

The Test in Prod team has actually implemented a minimal example in Python of what this looks like [here](https://github.com/testinprod-io/op-fast-sync). As you can see, they create a payload by querying the trusted L2 rpc, then use newPayload followed by forkchoiceUpdated to sync it.

To add further annoyance to this implementation, ethers-rs isn't able to to properly decode a block fetched from an optimism rpc due to a new tx type optimism adds. I do fortunately have a wip [branch of ethers](https://github.com/ncitron/ethers-rs/tree/optimism-support) that adds optimism support, although I haven't tested it yet. You should be able to switch our repo over to use this branch (or your own fork of it if you need to further patch it to get something working, and I can in parallel do some work to get this support upstreamed into ethers-rs (have already spoken to gakonst and he seems supportive of the effort).

Apologies for sending you off slightly in the wrong direction to start.

**Comment by ncitron** [comment]
Just pushed an update to make it an optional.

**Comment by merklefruit** [comment]
@ncitron Ok I had to make some other small changes to the RLP encoding (which is broken unfortunately) so I am currently using my fork of ethers-rs for this. Here's where I currently am:

- I ported the code from the repo you linked, using ethers to generate the RLP encoding of all the txs in the block
- I am sending the generated payload to op-geth
- The traces generated by op-geth upon request suggest the following error with decoding the deposited tx:

```
error="invalid transaction 0: rlp: expected input string or byte for common.Hash, 
decoding into (types.DepositTx).SourceHash"
```

For reference, you can find my ethers branch diffs [here](https://github.com/gakonst/ethers-rs/compare/master...merklefruit:ethers-rs:optimism-txs).

What I find most surprising is that the encoding done in `op-fast-sync` is different from what is described in the official specs [here](https://github.com/ethereum-optimism/optimism/blob/develop/specs/deposits.md#the-deposited-transaction-type), so I am having trouble debugging this.

Here's the RLP order according to the specs:
```
- sourceHash
- from
- to
- mint
- value
- data
- isSystemTx
- gasLimit
```

And here's the order used in `op-fast-sync`:
```
- sourceHash,
- from,
- to,
- mint,
- value,
- gas,
- isSystemTx,
- data
```

I've tried both orders without success for now.

My next idea is trying to produce the same payload with `op-fast-sync`, compare the two outputs and see what fields my current implementation is getting wrong. 

I appreciate any suggestion on this, as I feel I am a bit stuck

**Comment by ncitron** [comment]
Oh I see our problem. The RLP encoding of the `to` field is whacky for type 0x7E (deposited) transactions. Take a look at how we do the RLP encoding [here](https://github.com/a16z/magi/blob/master/src/derive/stages/attributes.rs#L222).

Note that I think you don't need to do the list stuff in the ethers version (might be wrong though so try both if it doesn't work)

**Comment by merklefruit** [comment]
I don't see any difference between the two implementations, the `rlp_opt` function does the same as the if...else block in magi's implementation, it seems

**Comment by merklefruit** [comment]
Oh wait, I think the tx type is encoded differently

**Comment by ncitron** [comment]
Oh yeah that's right. Did it have to do with not prepending the tx type for the new txs? I forgot that in my implementation.

**Comment by merklefruit** [comment]
Not only that, a couple more things were missing, I finally got it working after checking the RLP manually against the result from op-fast-sync 😄

**Comment by merklefruit** [comment]
In the past days I've looked into several eth2 consensus clients, lighthouse in particular, to find where and how they do checkpoint sync. Here's what I learned:

- the idea of setting a forkchoice and waiting for the execution client to sync from its p2p layer is right
- however simply setting a new `forkchoice` doesn't seem to be enough for geth, or I am doing something wrong

So my current problem is that when calling `forkchoiceUpdatedV1`, geth starts a sync cycle but it's returning false to `eth_syncing` requests. 

Here's my setup:

- op-geth running default command (verbosity 5 to debug)
- magi running with this branch and the following command: `cargo run --release -- --network optimism-goerli --jwt-secret bf54...015ff --l1-rpc-url https://eth-goerli...._o_Z --l2-rpc-url http://localhost:8545 --l2-engine-url http://localhost:8551 --l2-trusted-rpc-url https://opt-goerli...aw_o_Z --sync-mode checkpoint --checkpoint-hash 0xc2794a16acacd9f7670379ffd12b6968ff98e2a602f57d7d1f880220aa5a4973`

```text
op-geth  | DEBUG[04-29|13:16:33.505] FS scan times                            list="208.304µs" set="6.417µs"  diff="5.25µs"
op-geth  | TRACE[04-29|13:16:33.662] Engine API request received              method=NewPayload number=8,453,214 hash=c2794a..5a4973
op-geth  | TRACE[04-29|13:16:33.668] New skeleton head announced              number=8,453,214 hash=c2794a..5a4973 force=false
op-geth  | WARN [04-29|13:16:33.668] Ignoring payload with missing parent     number=8,453,214 hash=c2794a..5a4973 parent=183735..aa6b79
op-geth  | DEBUG[04-29|13:16:33.668] Served engine_newPayloadV1               conn=172.18.0.1:51016 reqid=1 duration=5.733298ms
op-geth  | TRACE[04-29|13:16:33.670] Engine API request received              method=ForkchoiceUpdated head=c2794a..5a4973 finalized=c2794a..5a4973 safe=c2794a..5a4973
op-geth  | INFO [04-29|13:16:33.670] Forkchoice requested sync to new head    number=8,453,214 hash=c2794a..5a4973
op-geth  | TRACE[04-29|13:16:33.670] New skeleton head announced              number=8,453,214 hash=c2794a..5a4973 force=true
op-geth  | DEBUG[04-29|13:16:33.670] Restarting skeleton subchain             head=8,453,214      tail=8,453,214
op-geth  | DEBUG[04-29|13:16:33.670] Dropping skeleton subchain               head=8,453,214      tail=8,453,214
op-geth  | DEBUG[04-29|13:16:33.670] Created new skeleton subchain            head=8,453,214      tail=8,453,214
op-geth  | DEBUG[04-29|13:16:33.670] Served engine_forkchoiceUpdatedV1        conn=172.18.0.1:51016 reqid=1 duration="588.618µs"
op-geth  | DEBUG[04-29|13:16:33.671] Starting reverse header sync cycle       head=8,453,214      hash=c2794a..5a4973 cont=8,453,213
op-geth  | DEBUG[04-29|13:16:33.673] Recalculated msgrate QoS values          proto=eth rtt=20s confidence=1.000 ttl=1m0s next=2023-04-29T13:16:53+0000
op-geth  | TRACE[04-29|13:16:33.673] Debug dump of mean capacities            proto=eth caps=map[]
op-geth  | DEBUG[04-29|13:16:33.678] Served eth_syncing                       conn=172.18.0.1:41562 reqid=1 duration="60.337µs"
op-geth  | DEBUG[04-29|13:16:36.511] FS scan times                            list="215.429µs" set="7.459µs"  diff="5.584µs"
```

**Comment by ncitron** [comment]
That's interesting that it returns false for syncing even though the logs clearly indicate that it is starting a reverse sync.

Does op-geth appear to be syncing correctly? If so, you can probably just use `eth_getBlockByNumber` with the `latest` blocktag to see whether it has synced up to the checkpoint block.

**Comment by merklefruit** [comment]
> Does op-geth appear to be syncing correctly? 

So I've tried both endpoints and let it run for a while, but no success so far. Will update when I make some progress.
Sorry if this is taking longer than expected!

**Comment by merklefruit** [comment]
By looking at op-erigon's docs, I found this line:

> --maxpeers=0, --nodiscover
[Optional] Disable P2P. Execution-layer peering is currently not supported in the Optimism protocol. Though this is not required, it saves resources since TX pool gossip is currently not available.

So, if execution layer peering is not supported (and we're running op-geth with the `nodiscover` flag) isn't this a problem for syncing?

**Comment by ncitron** [comment]
Yeah you were on the right track for the issue, and after debugging for a couple of hours I got op-geth to start syncing using this branch! Turns out there's a whole lot of weird stuff going on that makes it hard to find a peer to sync on optimism.

To start, we need to enable p2p networking. To do this we can just remove the `maxpeers` and `nodiscover` flags.

Now, when we try to sync op-geth actually tries to find peers, which is a good start. Unfortunately, all of the other versions of op-geth have peering turned off, so we need to provide our own peer to begin syncing with. I tried enabling p2p networking on one of my synced op-geth nodes on optimism goerli, but couldn't seem to get them to connect. It turns out the reason for this is because the enode (a unique identifier for peering) is determined by a value in `$DATADIR/geth/nodekey`, which has already been set in the original datadir that the Optimism team provides and we use in our docker images. This means when the unsynced node reached out to the synced node, it got confused since they had the same enode id.

To resolve this, I just deleted that file in my pre-synced op-geth node, which gave it a brand new enode key. Now anyone should be able to do checkpoint sync by using my node's enode.

To add a new peer, we need to enable geth's `admin` rpc namespace. To do that, we must modify the `http.api` flag in op-geth and add `admin` to the list. Once we do that, we can start up a geth node, and use the `admin_addPeer` method to add my synced node: 

`enode://e85ba0beec172b17f53b373b0ab72238754259aa39f1ae5290e3244e0120882f4cf95acd203661a27c8618b27ca014d4e193266cb3feae43655ed55358eedb06@3.86.143.120:30303?discport=21693`

So TLDR is that we need to just add the admin namespace and then add my synced node as a peer.

Longer term, I can host a couple of synced nodes with unique enode keys, and add them to the bootnode list to ensure everyone can find me.

I'm however now noticing that partway through syncing op-geth is crashing. This may just be a bug in op-geth that we are blocked by, but will keep looking into it.

**Comment by merklefruit** [comment]
> To start, we need to enable p2p networking. To do this we can just remove the maxpeers and nodiscover flags.

Exactly what I tried doing this morning, except that it couldn't find peers and kept throwing weird errors!

> so we need to provide our own peer to begin syncing with. 

That makes sense! I couldn't find any documentation on this... I had started asking around in discords for help 😄 

> To resolve this, I just deleted that file in my pre-synced op-geth node, which gave it a brand new enode key. Now anyone should be able to do checkpoint sync by using my node's enode.

I'm gonna try with your enode, at least to polish up this branch a little bit

**Comment by merklefruit** [comment]
Ok it's currently syncing thanks to your enode!

I still find it weird that `eth_syncing` is returning false, and `eth_blockNumber` is still not advancing for some reason.
I'll give it some more time until it either crashes or syncs completely

File: src/runner/mod.rs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        if l2_provider.get_block(checkpoint_hash).await?.is_some() {
+            tracing::warn!("finalized head is above the checkpoint block");
+            self.start_driver().await?;
+            return Ok(());
+        }
+
+        // this is a temporary fix to allow execution layer peering to work
+        // TODO: use a list of whitelisted bootnodes instead
+        tracing::info!("adding trusted peer to the execution layer");
+        l2_provider.add_peer(TRUSTED_PEER_ENODE.to_string()).await?;
```
Comment (merklefruit) [inline src/runner/mod.rs:160]: Should I leave this in? It's probably not needed now as you've added op-geth nodes with the right config

File: src/driver/mod.rs [hunk 2]
```diff
@@ -43,30 +43,29 @@ pub struct Driver<E: Engine> {
...
-            .flatten()
-            .and_then(|block| block.try_into().ok());
-
-        let finalized_head = head
-            .as_ref()
-            .map(|h| h.l2_block_info)
-            .unwrap_or(config.chain.l2_genesis);
+        let block_id = BlockId::Number(BlockNumber::Finalized);
+        let head = match HeadInfo::from_block(block_id, &provider).await? {
+            Some(head) => head,
```
Comment (merklefruit) [inline src/driver/mod.rs:55]: I have one doubt on this line: should we start from the execution layer `finalized` head after syncing?
I've noticed that if we do this after a full checkpoint sync cycle, it starts from the genesis block, wasting time.

File: src/runner/mod.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        self.sync_mode = sync_mode;
+        self
+    }
+
+    pub fn with_checkpoint_hash(mut self, checkpoint_hash: Option<String>) -> Self {
+        self.checkpoint_hash = checkpoint_hash;
+        self
+    }
+
+    pub async fn run(&self) -> Result<()> {
```
Comment (ncitron) [inline src/runner/mod.rs:61]: Can we make this consume self (take `self` instead of `&self`) to ensure that run is only called once?

File: src/runner/mod.rs (hunk 1 above)
Comment (ncitron) [reply src/runner/mod.rs:147]: I think we should leave this for now. We can remove once there are more nodes that have peering enabled.

File: src/driver/types.rs [hunk 4]
```diff
@@ -13,6 +16,25 @@ pub struct HeadInfo {
     pub l1_epoch: Epoch,
 }
 
+impl HeadInfo {
```
Comment (ncitron) [inline src/driver/types.rs:19]: Instead of fetching the block and doing the conversion here, can we just implement `From<Block<Transaction>>` for `HeadInfo`? We can fetch the blocks as needed in the runner, and leave the `types` submodule focussed on types and type conversions.

File: src/engine/payload.rs [hunk 5]
```diff
@@ -37,6 +44,70 @@ pub struct ExecutionPayload {
...
+            .collect::<Vec<_>>())
+        .to_vec();
+
+        Ok(ExecutionPayload {
+            parent_hash: l2_block.parent_hash,
+            fee_recipient: SystemAccounts::default().fee_vault,
+            state_root: l2_block.state_root,
+            receipts_root: l2_block.receipts_root,
+            logs_bloom: l2_block.logs_bloom.unwrap().as_bytes().to_vec().into(),
+            prev_randao: l1_block.mix_hash.unwrap(),
```
[the remark this answers is not in the document]
Comment (ncitron) [reply src/engine/payload.rs:93]: Also if we do this, we can do a similar thing as I suggested for `HeadInfo` and implement `From<Block<Transactions>>` for `ExecutionPayload` which I think keeps things a bit cleaner with seperating type work from rpc calls.

File: bin/magi.rs [hunk 6]
```diff
@@ -68,13 +51,17 @@ pub struct Cli {
...
+    #[clap(short = 'p', long)]
+    rpc_port: Option<u16>,
     #[clap(long)]
     logs_dir: Option<String>,
     #[clap(long)]
     logs_rotation: Option<String>,
+    #[clap(long)]
+    checkpoint_hash: Option<String>,
+    #[clap(long)]
+    l2_trusted_rpc_url: Option<String>,
```
Comment (ncitron) [inline bin/magi.rs:64]: Can we rename this to `checkpoint_sync_url`? Same with the config and docker setup.

Usually consensus clients follow that naming convention, and I also think it helps with making it clear that this flag is only needed when when using checkpoint sync.

**Review by ncitron — COMMENTED** [review]
This is looking really good. Just left a few notes.

File: src/engine/api.rs [hunk 7]
```diff
@@ -155,6 +155,11 @@ impl EngineApi {
         // This scenario shouldn't occur as the response should always have either data or an error
         eyre::bail!("Failed to parse Engine API response")
     }
+
+    /// Calls the engine to verify it's available to receive requests
+    pub async fn is_available(&self) -> bool {
+        self.post::<Value>("eth_chainId", vec![]).await.is_ok()
```
[the remark this answers is not in the document]
Comment (merklefruit) [reply src/engine/api.rs:161]: I was looking at the engine api specs here https://github.com/ethereum/execution-apis/blob/main/src/engine/common.md#underlying-protocol

It says that a few methods including eth_chainId have to be available also on that port.
I can see op-geth processing this request in the logs, but if you prefer another solution I can set it to something different?

File: src/runner/mod.rs (hunk 1 above)
Comment (merklefruit) [reply src/runner/mod.rs:147]: Cool!

File: src/engine/api.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ncitron) [reply src/engine/api.rs:161]: Oh nevermind. I didn't know that. No need to make this change.

File: src/config/mod.rs [hunk 8]
```diff
@@ -131,6 +140,8 @@ pub struct ChainConfig {
     pub blocktime: u64,
     /// L2 To L1 Message passer address
     pub l2_to_l1_message_passer: Address,
+    /// The L1 block contract on L2
+    pub l1_block: Address,
```
[the remark this answers is not in the document]
Comment (merklefruit) [reply src/config/mod.rs]: Yep, I noticed this one and fee_vault as well were duplicated and have already removed them

File: bin/magi.rs (hunk 6 above)
Comment (merklefruit) [reply bin/magi.rs:64]: sure

File: src/engine/payload.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (merklefruit) [reply src/engine/payload.rs:93]: Looks like we can! Yep, this cleans things a lot 😄

File: src/driver/mod.rs (hunk 2 above)
Comment (merklefruit) [reply src/driver/mod.rs]: Yes, that's why I was surprised when it started doing this. I think it may have failed fetching that block and it fell back to that. I'm investigating

File: docker/start-magi.sh [hunk 9]
```diff
@@ -7,7 +7,7 @@ exec magi \
     --l1-rpc-url $L1_RPC_URL \
     --l2-rpc-url http://${EXECUTION_CLIENT}:8545 \
     --l2-engine-url http://${EXECUTION_CLIENT}:8551 \
-    --l2-trusted-rpc-url $TRUSTED_L2_RPC_URL \
     --rpc-port $RPC_PORT \
     --sync-mode $SYNC_MODE \
+    --checkpoint-sync-url $TRUSTED_L2_RPC_URL \
```
Comment (ncitron) [inline docker/start-magi.sh:12]: Shouldn't this be `$CHECKPOINT_SYNC_URL`?

**Review by ncitron — COMMENTED** [review]
One last note here, then I think it looks good to go, pending the one question about restarting from genesis.

File: docker/start-magi.sh (hunk 9 above)
Comment (merklefruit) [reply docker/start-magi.sh:12]: oops, good catch

**Comment by merklefruit** [comment]
I guess another thing I didn't mention is that I didn't test op-erigon support. In theory it should work out of the box, right? 

If so, should I remove the `--nodiscover` flag on its startup script?

File: src/driver/mod.rs (hunk 2 above)
Comment (merklefruit) [reply src/driver/mod.rs]: I couldn't reproduce it, right now it's picking up the right block
